File: bifrost-py/src/bifrost/server/default_methods.py
"""Default RPC methods auto-registered on every server.

Mirrors ``bifrost-ts/src/server/methods.ts`` and
``bifrost-ts/src/server/default-methods.ts``:

* ``rpc:on``  — subscribe to events on an optional channel
* ``rpc:off`` — unsubscribe
* ``rpc:logout`` — clear auth state on the node

``rpc:login`` and ``list:methods`` are NOT auto-registered (see
PROTOCOL.md §7.4, §7.5).
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from bifrost.protocol.constants import NO_CHANNEL, Methods, ServerEvents
from bifrost.server.method import Method, MethodOptions
from bifrost.server.room_name import get_room_name

if TYPE_CHECKING:
    from bifrost.server.client_node import ClientNode
    from bifrost.server.server import Server


async def _rpc_on(server: Server, node: ClientNode | None, params: Any) -> dict[str, bool]:
    if node is None:
        return {}

    events = _events_list(params)
    channel = _channel(params)

    if not events:
        return {}

    channel_allowed = await _resolve_channel_permission(server, node, channel)

    result: dict[str, bool] = {}
    for event_name in events:
        if not channel_allowed:
            result[event_name] = False
            continue
        result[event_name] = await _can_subscribe(server, node, event_name, channel)
    return result


async def _rpc_off(server: Server, node: ClientNode | None, params: Any) -> dict[str, bool]:
    if node is None or node.socket is None:
        return {}

    transport = server.websocket_transport
    if transport is None:
        return {}

    events = _events_list(params)
    channel = _channel(params)

    result: dict[str, bool] = {}
    for event_name in events:
        event = server.events.get(event_name)
        if event is None:
            result[event_name] = False
            continue
        transport.rooms.leave(node.socket, get_room_name(channel, event_name))
        result[event_name] = True
    return result


async def _rpc_logout(server: Server, node: ClientNode | None, _params: Any) -> bool:
    if node is None:
        return False
    node.context = {}
    node.authenticated = False
    node.user_id = None
    server.emit_server_event(ServerEvents.LOGOUT, node)
    return True


async def _can_subscribe(
    server: Server, node: ClientNode, event_name: str, channel: str
) -> bool:
    event = server.events.get(event_name)
    if event is None:
        return False
    if event.is_protected and not node.authenticated:
        return False

    allowed = await event.evaluate_should_subscribe(node, event_name, channel)
    if not allowed:
        return False

    if node.socket is None:
        return False

    transport = server.websocket_transport
    if transport is None:
        return False

    transport.rooms.join(node.socket, get_room_name(channel, event_name))
    return True


async def _resolve_channel_permission(
    server: Server, node: ClientNode, channel: str
) -> bool:
    import inspect as _inspect

    result = server.should_allow_channel_subscribe(node, channel)
    if _inspect.isawaitable(result):
        result = await result
    return bool(result)


def _events_list(params: Any) -> list[str]:
    if isinstance(params, dict):
        events = params.get("events")
        if isinstance(events, list):
            return [e for e in events if isinstance(e, str)]
    return []


def _channel(params: Any) -> str:
    if isinstance(params, dict):
        value = params.get("channel", NO_CHANNEL)
        if isinstance(value, str) and value:
            return value
    return NO_CHANNEL


def build_default_methods(server: Server) -> dict[str, Method]:
    """Return the auto-registered method set for a fresh server."""

    return {
        Methods.RPC_ON.value: Method(
            server,
            Methods.RPC_ON.value,
            lambda node, params: _rpc_on(server, node, params),
            MethodOptions(protected=False),
        ),
        Methods.RPC_OFF.value: Method(
            server,
            Methods.RPC_OFF.value,
            lambda node, params: _rpc_off(server, node, params),
            MethodOptions(protected=False),
        ),
        Methods.RPC_LOGOUT.value: Method(
            server,
            Methods.RPC_LOGOUT.value,
            lambda node, params: _rpc_logout(server, node, params),
            MethodOptions(protected=True),
        ),
    }

File: bifrost-rs/crates/bifrost-runtime/src/default_methods.rs
//! Default RPC methods auto-registered on every `Server` — port of
//! `bifrost-py/src/bifrost/server/default_methods.py`.
//!
//! Ships:
//!
//! * `rpc:on`  — subscribe to events on an optional channel
//! * `rpc:off` — unsubscribe
//! * `rpc:logout` — clear auth state on the node
//!
//! `rpc:login` and `list:methods` are NOT auto-registered (PROTOCOL.md
//! §7.4, §7.5). The application registers `rpc:login` itself when it
//! opts into auth.

use crate::client_node::ClientNode;
use crate::method::{BoxResult, MethodOptions, RpcHandler};
use crate::server::Server;
use bifrost_protocol::{NO_CHANNEL, methods, room_name};
use futures::FutureExt as _;
use serde_json::{Map, Value, json};
use std::sync::Arc;

/// Install `rpc:on` / `rpc:off` / `rpc:logout` on `server`.
pub(crate) fn register(server: &Arc<Server>) {
    server.add_method(
        methods::RPC_ON,
        rpc_on(server.clone()),
        MethodOptions::default(),
    );
    server.add_method(
        methods::RPC_OFF,
        rpc_off(server.clone()),
        MethodOptions::default(),
    );
    server.add_method(
        methods::RPC_LOGOUT,
        rpc_logout(server.clone()),
        MethodOptions {
            protected: true,
            ..Default::default()
        },
    );
}

/// `rpc:on` — subscribe a connected client to a list of events on a
/// channel. Returns `{ event_name: bool }` per-event success.
fn rpc_on(server: Arc<Server>) -> RpcHandler {
    Arc::new(move |node: Arc<ClientNode>, params: Value| {
        let server = server.clone();
        async move {
            let events = events_list(&params);
            let channel = channel_or_default(&params);

            if events.is_empty() {
                return Ok(Value::Object(Map::new())) as BoxResult;
            }

            let mut result = Map::new();
            for event_name in events {
                let allowed = can_subscribe(&server, &node, &event_name, &channel).await;
                result.insert(event_name, Value::Bool(allowed));
            }
            Ok(Value::Object(result))
        }
        .boxed()
    })
}

/// `rpc:off` — unsubscribe from the same shape `rpc:on` accepts.
fn rpc_off(server: Arc<Server>) -> RpcHandler {
    Arc::new(move |node: Arc<ClientNode>, params: Value| {
        let server = server.clone();
        async move {
            let events = events_list(&params);
            let channel = channel_or_default(&params);

            // No socket → nothing to unsubscribe; return an empty map.
            let Some(socket) = node.socket.as_ref().cloned() else {
                return Ok(Value::Object(Map::new())) as BoxResult;
            };
            let Some(rooms) = server.rooms.read().ok().and_then(|opt| opt.clone())
            else {
                return Ok(Value::Object(Map::new())) as BoxResult;
            };

            let mut result = Map::new();
            for event_name in events {
                let exists = server
                    .events
                    .read()
                    .map(|map| map.contains_key(&event_name))
                    .unwrap_or(false);
                if !exists {
                    result.insert(event_name, Value::Bool(false));
                    continue;
                }
                rooms.leave(&*socket, &room_name(&channel, &event_name));
                result.insert(event_name, Value::Bool(true));
            }
            Ok(Value::Object(result))
        }
        .boxed()
    })
}

/// `rpc:logout` — clear auth state on the node. Always returns `true`.
fn rpc_logout(_server: Arc<Server>) -> RpcHandler {
    Arc::new(move |node: Arc<ClientNode>, _params: Value| {
        async move {
            node.set_authenticated(false);
            *node.context.write().expect("context poisoned") = Value::Null;
            *node.user_id.write().expect("user_id poisoned") = None;
            Ok(json!(true)) as BoxResult
        }
        .boxed()
    })
}

// ---------------------------------------------------------------------------
// helpers
// ---------------------------------------------------------------------------

fn events_list(params: &Value) -> Vec<String> {
    params
        .get("events")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default()
}

fn channel_or_default(params: &Value) -> String {
    params
        .get("channel")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .unwrap_or(NO_CHANNEL)
        .to_string()
}

async fn can_subscribe(
    server: &Server,
    node: &Arc<ClientNode>,
    event_name: &str,
    channel: &str,
) -> bool {
    let event = match server.events.read().ok().and_then(|m| m.get(event_name).cloned()) {
        Some(e) => e,
        None => return false,
    };
    if event.is_protected && !node.is_authenticated() {
        return false;
    }
    let allowed = (event.should_subscribe)(
        node.clone(),
        event_name.to_string(),
        channel.to_string(),
    )
    .await;
    if !allowed {
        return false;
    }
    let Some(socket) = node.socket.as_ref().cloned() else {
        return false;
    };
    let Some(rooms) = server.rooms.read().ok().and_then(|opt| opt.clone()) else {
        return false;
    };
    rooms.join(socket, room_name(channel, event_name));
    true
}
